Sources/Hummingbird/Middleware/TracingMiddleware.swift

```diff
@@ -0,0 +1,136 @@
+/// You may opt in to recording a specific subset of HTTP request/response header values by passing
+/// a set of header names to ``init(recordingHeaders:)``.
+@available(macOS 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *)
+public struct HBTracingMiddleware: HBMiddleware {
+    private let headerNamesToRecord: Set<RecordingHeader>
+
+    /// Intialize a new HBTracingMiddleware.
+    ///
+    /// - Parameter recordingHeaders: A list of HTTP header names to be recorded as span attributes. By default, no headers
+    /// are being recorded.
+    public init<C: Collection>(recordingHeaders headerNamesToRecord: C) where C.Element == String {
+        self.headerNamesToRecord = Set(headerNamesToRecord.map(RecordingHeader.init))
+    }
+
+    /// Intialize a new HBTracingMiddleware.
+    public init() {
+        self.init(recordingHeaders: [])
+    }
+
+    public func apply(to request: HBRequest, next: HBResponder) -> EventLoopFuture<HBResponse> {
+        var baggage = request.baggage
+        InstrumentationSystem.instrument.extract(request.headers, into: &baggage, using: HTTPHeadersExtractor())
+
+        let operationName: String = {
+            guard let endpointPath = request.endpointPath else {
+                return "HTTP \(request.method.rawValue) route not found"
+            }
+            return endpointPath
+        }()
+
+        return request.withSpan(operationName, baggage: baggage, ofKind: .server) { request, span in
+            span.attributes["http.method"] = request.method.rawValue
+            span.attributes["http.target"] = request.uri.path
+            span.attributes["http.flavor"] = "\(request.version.major).\(request.version.minor)"
+            span.attributes["http.scheme"] = request.uri.scheme?.rawValue
+            span.attributes["http.user_agent"] = request.headers.first(name: "user-agent")
+            span.attributes["http.request_content_length"] = request.body.buffer?.readableBytes
+
+            span.attributes["net.host.name"] = request.application.server.configuration.address.host
+            span.attributes["net.host.port"] = request.application.server.configuration.address.port
+
+            if let remoteAddress = request.remoteAddress {
+                span.attributes["net.sock.peer.port"] = remoteAddress.port
+
+                switch remoteAddress.protocol {
+                case .inet:
+                    span.attributes["net.sock.peer.addr"] = remoteAddress.ipAddress
+                case .inet6:
+                    span.attributes["net.sock.family"] = "inet6"
+                    span.attributes["net.sock.peer.addr"] = remoteAddress.ipAddress
+                case .unix:
+                    span.attributes["net.sock.family"] = "unix"
+                    span.attributes["net.sock.peer.addr"] = remoteAddress.pathname
+                default:
+                    break
+                }
+            }
+
+            recordHeaders(request.headers, toSpan: span, withPrefix: "http.request.header.")
```

Comment:
will this include the authorization header in the trace?

Reply:
By default, nothing is recorded. If you'd choose to include the `Authorization` header, yes, it would be recorded as `http.request.header.authorization`.

Reply:
Thanks, I misread the code. I know see it is only recording what you pass into the initializer